src/graph_compute/concurrent_queue.h

```diff
@@ -1,48 +1,54 @@
 #ifndef CONCURRENT_QUEUE_H_
 #define CONCURRENT_QUEUE_H_
 
+#include <iostream>
 #include "../libgraphio/libgraphio.h"
 
+using namespace std;
+
 struct mrmw_queue_t {
-  volatile vid_t * data;
-  vid_t sentinel;
+  volatile vid_t * const data;
+  const vid_t sentinel;
   size_t mask;
-  size_t head;
-  size_t tail;
-  mrmw_queue_t(volatile vid_t * _data, size_t _numBits, vid_t _sentinel = -1) {
-    data = _data;
-    mask = (1 << _numBits) - 1;
-    tail = 0;
-    head = 0;
-    sentinel = _sentinel;
-    for (size_t i = 0; i <= mask; i++) {
-      data[i] = sentinel;
-    }
-  }
+  volatile size_t head;
+  volatile size_t tail;
+  volatile int lock;
+  mrmw_queue_t(volatile vid_t * const _data,
+               size_t _numBits,
+               const size_t _sentinel = static_cast<size_t>(-1)) :
+               data(_data),
+               sentinel(_sentinel),
+               mask((1 << _numBits) - 1),
+               head(0),
+               tail(0),
+               lock(0) { }
   // assumes that it is not possible to overflow
-  void push(vid_t value);
-  vid_t pop();
+  void push(const vid_t value);  //  So, push always succeeds
+  vid_t pop();  //  pop can return sentinel, if it is full
 };
 typedef struct mrmw_queue_t mrmw_queue_t;
 
-void mrmw_queue_t::push(vid_t value) {
-  size_t position = __sync_fetch_and_add(&tail, 1);
-  while (data[position & mask] != sentinel) {}
-  data[position & mask] = value;
-  __sync_synchronize();
+inline void mrmw_queue_t::push(const vid_t value) {
+  while (__sync_lock_test_and_set(&lock, 1) == 1) {}
```

Comment:
The lock-free (which wasn't technically lock-free, since a thread could be preempted prior to writing the sentinel or new value, for pop and push, respectively) approach seemed to have an issue when I allowed multiple readers.  I was doing an atomic and a store-queue flush with the old version, the same as the locking version.  The only theoretical advantage of the previous version was that readers and writers could operate simultaneously.  Nonetheless, correct slow code is better than fast broken code.